src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -383,7 +383,96 @@ protected void InitializeReferences()
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                return runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                bool skipScan = false;
+                runner.InitializeTimeout(internalMatchTimeout);
+                ReadOnlySpan<char> span = input.AsSpan(beginning, length);
```

Comment:
I don't understand how this is avoiding allocations when one of the string-based entrypoints is used with an existing CompileToAssembly assembly.  Doesn't it need to store the `input` into `runtext` so that the original string is available to overrides of `FindFirstChar` and `Go`?

Reply:
A valid alternative to `if (inputSpan != runtext) runtext = inputSpan.ToString();` would also be to throw an exception in the base Scan method if runtext is null.  That should only be hit if a) a new span-based entrypoint is used and b) an existing CompileToAssembly assembly is being used that doesn't override Scan, and we could say that combination simply isn't supported.

Reply:
The way I have it now, is that instead of `Regex.Run` doing the `runner.runtext = input`, I'm doing that in the default implementation of Scan as you suggest:

https://github.com/dotnet/runtime/blob/cc926232aa022da65d932907ab6628439369f922/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs#L91-L103

This way, setting runtext will ONLY happen now when using CompiledToAssembly. All of the other engines will override the Scan method, and won't be setting runtext any longer, which is how we prevent that extra allocation.

Reply:
> The way I have it now, is that instead of Regex.Run doing the runner.runtext = input, I'm doing that in the default implementation of Scan as you suggest:

We need to do _both_.  If you don't set `runner.runtext = input` in the `Run` method, then when it gets to this point in `Scan`, `runtext` will be null and we'll always end up doing text.ToString.